Documentation: Vault Docs (#8936)
thepracticaldev/dev.to: docs/backend/configuration.md

# Configuration

> We currently use three gems for configuring the application:

Reviewer: I'd remove the number so that we don't need to change it if we increase or decrease the number of gems :D 

"We currently use a few gems"

Author: "We currently use the following gems..." :)

[Accessing the admin panel](./admin.md)).

![Screenshot of site configuration admin interface](https://user-images.githubusercontent.com/47985/73627238-6276d500-467e-11ea-8724-afb703f056bc.png)

## Vault

The vault Ruby gem allows us to interact with

> [HashiCorp's Vault](https://www.vaultproject.io/docs/what-is-vault). In a

Reviewer 2: ```suggestion
[Vault](https://www.vaultproject.io/docs/what-is-vault). In a
```

Minor, but `The vault Ruby gem allows us to interact with HashiCorp's Vault` reads to me like "it's a gem that lets us interact with _their_ vault". If we just ✂️ the `HashiCorp's` here, that potential confusion will be removed.

nutshell, Vault is a tool for securely storing and accessing secrets. The way we

> have it configured, it is completely optional for running a Forem. To access it

Reviewer 2: I think we could just cut `The way we have it configured`, since that doesn't seem particularly relevant to these docs because we don't ever tell anyone how to "un-configure" it :)

we use the wrapper `AppSecrets`.

```ruby
class AppSecrets
  def self.[](key)
    result = Vault.kv(namespace).read(key)&.data&.fetch(:value) if ENV["VAULT_TOKEN"].present?
    result ||= ApplicationConfig[key]

    result
  rescue Vault::VaultError
    ApplicationConfig[key]
  end

  def self.[]=(key, value)
    Vault.kv(namespace).write(key, value: value)
  end

  def self.namespace
    ENV["VAULT_SECRET_NAMESPACE"]
  end
  private_class_method :namespace
end
```

> As you can see in the code, we attempt to access a secret from Vault if it is

Reviewer 2: ```suggestion
We attempt to access a secret from Vault if it is
```

Might not be evident to someone who is more of a beginner, and they might read "as you can see" and wonder why they don't see. If you wanted to keep it in, I'd recommend pointing to the line where this is happening to make it more clear. But, I think it's also fine to just ✂️ it, too.

> enabled, ie the `VAULT_TOKEN` is present. If Vault is not enabled or we cannot

Reviewer 2: ```suggestion
enabled, i.e. if the `VAULT_TOKEN` is present. If Vault is not enabled or if we cannot
```

> find the secret in it, then we fallback on fetching the secret from the

Reviewer 2: ```suggestion
find the secret in it, then we fallback to fetching the secret from the
```

`ApplicationConfig`.

> One advantage to using Vault with Forem is that it allows you to update your

Reviewer 2: ```suggestion
One advantage of using Vault with Forem is that it allows you to update your
```

secrets easily through the application rather than having to mess with ENV

> files. If you would like to tryout Vault follow our

Reviewer 2: ```suggestion
files. If you would like to try out Vault, follow our
```

> [installation guide for setting it up locally](installation/vault).

Reviewer 2: ```suggestion
[installation guide for setting it up locally](../installation/vault.md).
```
Does this path actually link correctly without a `../` before it? I'd suspect you need to change it to this since `/installation` and `backend` are in separate directories.

Author: Yeah I am pretty sure this will work. If you look at the installation README we have similar links
```
## Installing locally

You can install DEV to your local machine and we have instructions for
[Mac](/installation/mac), [Windows](/installation/windows) and
[Linux](/installation/linux).
```
